withdrawals as part of the coin selection

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1861 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 12b28f75b69c4bef47e1d1b7f2b765e9d544f800
  :round_pushpin: **take withdrawals into account, one level earlier, during the coin selection**
  Still to be done:

    - Make sure it's correctly done in the largest-first algorithm
    - Add some test scenario that show the influence of the withdrawal

- 547b1301fe7010cc0285a651aa8448cf88b6abc7
  :round_pushpin: **unit-test the newly introduced 'proportionallyTo'**
  Getting this one wrong would be quite bad :s

- 975f72648879a4f9ee513592e322dd10565f5739
  :round_pushpin: **add unit tests showing how withdrawal impacts the random coin selection**


> :bulb: NOTE
>
> I've chosen not to treat the withdrawal as a single _input_, but more as a "money pot" that is proportionally distributed amongst change output, so that it contributes to every output, based on their size. This is to avoid having a small output consuming the entire withdrawal for itself. Note sure if I'll keep the approach in the end, I'll have the night to think about it.

# Comments

<!-- Additional comments or screenshots to attach if any -->

- [ ] TODO: take into account the withdrawal when doing largest-first (and testing it)
- [ ] TODO: echo the error message change on "ErrInputsDepleted" to integration tests relying on that message.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
take withdrawals into account, one level earlier, during the coin selection

Still to be done:

- Make sure it's correctly done in the largest-first algorithm
- Add some test scenario that show the influence of the withdrawal

KtorZ committed:
unit-test the newly introduced 'proportionallyTo'

Getting this one wrong would be quite bad :s

KtorZ committed:
add unit tests showing how withdrawal impacts the random coin selection

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1878,10 +1878,9 @@ instance Buildable e => LiftHandler (ErrCoinSelection e) where
-                , "Try sending a smaller amount."
+                [ "I cannot select enough UTxO from your wallet to construct "
+                , "an adequate transaction. Try sending a smaller amount or "
+                , "increasing the number of available UTxO."
```
`ErrInputsDepleted` is now used also when trying to select coins on a null UTxO, despite enough balance in the reward account.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs:
```diff
@@ -112,64 +114,89 @@ random
+            (\acc out ->
+                let
+                    withdraw = totalWithdraw
+                        `proportionallyTo` (getCoin (coin out) % totalOut)
```
So here, we are guaranteed that the ratio given as second argument is always `<= 1`

paweljakubas replied:
real question, maybe trivial... but why do we need to use `proportionallyTo` at all? why this rule to transfer upon withdrawing proportionally to relative weight in total outs?  Why not to transfer randomly (adding up to withdrawal)? Or almost everything to one plus 1 + 1 + ... + 1? Or evenly? Or create one change?

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs:
```diff
@@ -112,64 +114,89 @@ random
-            guard sel $> (sel, remUtxo)
+            -- NOTE re-assigning total withdrawal to cope with potential
+            -- rounding issues.
+            guard sel $> (sel { withdrawal = totalWithdraw }, remUtxo)
```
`sel` here is constructed by appending multiple selection together from left to right, using the `Monoid` instance on `CoinSelection`. So in the end, the total withdrawal that we get is the sum of all withdrawals that were assigned to each output, but because `proportionallyTo` rounds **down**, the final sum may end up being slightly less than the total (an error of 1 per output at most) yet, we need it to be _precisely equal_ to the reward's balance without what the Ledger will reject the withdrawal. 

Thus, this is reassigned here, and the possible exceeding will be balanced out during fee balancing anyway.

paweljakubas replied:
if we transfer withdraw to one change we would not have this problem;-) But, when used `proportionallyTo` indeed we need to take care of this... So what happens in detail when  `proportionallyTo` rounding does consume not all withdraw? To which changes this additional `1`s go ? To those that correspond to outputs that were affected by this rounding, right?

KtorZ replied:
> To which changes this additional 1s go

Most likely, they'll be used to partially cover fee in the fee balancing algorithm here. What will happen in case of rounding issues is that the transaction will be _slightly_ unbalanced (positively) and have a few lovelace extra on the input side assigned to "nothing". Following the fee balancing algorithm, it'll start by consuming that delta, and then, starts to deplete (proportionally again) change outputs. So in the end, the little exceeding isn't lost but goes directly for fee; which is roughly similar to assigning it to any of the change output, since they'll be depleted for fees regardless.

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:
```diff
@@ -117,12 +123,35 @@ changeBalance = foldl' addCoin 0 . change
 
+-- | Total UTxO balance + withdrawal.
+totalBalance :: Quantity "lovelace" Word64 -> [(TxIn, TxOut)] -> Word64
+totalBalance (Quantity withdraw) inps = balance' inps + withdraw
```
:+1:

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:
```diff
@@ -117,12 +123,35 @@ changeBalance = foldl' addCoin 0 . change
+--
+-- >>> 10 `proportionallyTo` 1%3
+-- 3
+proportionallyTo :: Integral a => a -> Ratio a -> a
```
any chance to have unit test for this or something that uses this inside?

paweljakubas replied:
it is checked below

KtorZ replied:
Indeed!

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/LargestFirst.hs:
```diff
@@ -59,11 +69,15 @@ largestFirst opt outs utxo = do
             let moneyRequested = sum $ (getCoin . coin) <$> (descending outs)
-            let utxoBalance = fromIntegral $ balance utxo
+            let utxoList = Map.toList $ getUTxO utxo
+            let total = totalBalance withdraw utxoList
```
:+1:

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:
```diff
@@ -151,15 +160,106 @@ spec = do
+        coinSelectionUnitTest random "withdrawal simple"
+            (Right $ CoinSelectionResult
+                { rsInputs = [1]
+                , rsChange = []
```
why it is not `[1]` ?

paweljakubas replied:
ah, ok. we have input 1 plus withdraw 1 to cover 2

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:
```diff
@@ -151,15 +160,106 @@ spec = do
+        coinSelectionUnitTest random "withdrawal multi-output"
+            (Right $ CoinSelectionResult
+                { rsInputs = [1,1]
+                , rsChange = []
```
why it is not `[1,1]` ?

paweljakubas replied:
inps  + withdrawals : `[1,1,2]` to  cover `[2,2]` gives `[]` change

KtorZ replied:
These examples are indeed a bit tricky to follow, I tried to left some note above each test case to make it a bit easier.

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:
```diff
@@ -151,15 +160,106 @@ spec = do
+                , totalWithdrawal = 20
+                })
+
+        coinSelectionUnitTest random "withdrawal requires at least one input"
```
this one I understand :+1:

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:
```diff
@@ -151,15 +160,106 @@ spec = do
+            -- - 2 Ada goes to the other output ( 4/14 * 10 ~= 2)
+            (Right $ CoinSelectionResult
+                { rsInputs = [5,5]
+                , rsChange = [2,3]
```
and here we see that in spite of missing `1` by proportionallyTo the money is counting right

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:
```diff
@@ -151,15 +160,106 @@ spec = do
+                { maxNumOfInputs = 100
+                , validateSelection = noValidation
+                , utxoInputs = [1]
+                , txOutputs = 10 :| [10]
```
in old days it would trigger not enough fragmentation if the number of inputs does not match number of outputs. now I see withdrawals help in this sense :+1:  wonder if there are some integration tests checking this edge conditions that needs to be updated

paweljakubas replied:
what about this example : 
```
            (CoinSelectionFixture
                { maxNumOfInputs = 100
                , validateSelection = noValidation
                , utxoInputs = [1]
                , txOutputs = 1 :| [1,1,1,1,1]
                , totalWithdrawal = 5
                })
```
?

KtorZ replied:
> what about this example

Following the _current_ logic, each output would receive `0` withdrawal bonus (1/6 * 5, rounded down) and therefore, this would fail on the second output because there wouldn't be enough inputs to select. Although indeed, if instead of using the withdrawal proportionally, we were using them fully to cover "just" the right amount, it'd give better result.

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelectionSpec.hs:
```diff
@@ -99,6 +109,16 @@ spec = do
         prop "All inputs are used per transaction" prop_allInputsAreUsedPerTx
         prop "Addresses are recycled fairly" prop_fairAddressesRecycled
+
+    describe "proportionallyTo" $ do
```
ah, this one I was asking above - persuasive :100:

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs:
```diff
@@ -707,10 +709,10 @@ genSelectionFor :: NonEmpty TxOut -> Gen CoinSelection
     utxo <- vector (NE.length outs * 3) >>= genUTxO
-    case runIdentity $ runExceptT $ largestFirst opts outs utxo of
+    withdrawal_ <- genWithdrawal
+    case runIdentity $ runExceptT $ largestFirst opts outs (Quantity withdrawal_) utxo of
```
Dusting is going to affect this somehow ? Or we are confident it is going to be fine?

KtorZ replied:
I am not sure to follow your concern on this one :thinking:  ?

KtorZ committed:
consume withdrawal gradually, instead of spliting them proportionally

It is a bit more work to get this right, but in the end, it solves the same problem (being able to use the withdrawal on several outputs) with more elegance and efficiency.
Now the withdrawal is used gradually, and only what's needed at every step so in that sense, it is optimal.

KtorZ committed:
adjust test scenarios to show how the coin selection now gradually use the withdrawal

KtorZ committed:
have largest-first treat outputs as a whole instead of one-by-one

  The Largest-First algorithm is used as a fall-back for cases when more sophisticated algorithms fail to produce a coin selection.
  However, our current implementation is broken in a subtle way. Rather than considering the total collective value of all outputs when attempting to find a selection of inputs (as it should), our implementation instead considers each output independently, and for each output, attempts to find a selection of inputs to cover just that output.
  This leads to an unnecessary cardinality restriction, namely that a single unique input can only be used to pay for at most one unique output, which means that any set of n outputs can only be paid for if the UTxO has at least n inputs. Under this restriction, the algorithm can fail even if the total value of inputs is sufficient to cover the required outputs.

KtorZ committed:
add a few scenarios covering withdrawals for largest-first

paweljakubas reviewed (approved):
LGTM!

KtorZ committed:
remove now obsolete integration tests

The server no longer fails when the number of inputs is smaller than the number of outputs, the fallback now consider the output list as a whole
